File: app/src/main/java/dev/ragnarok/fenrir/mvp/presenter/RepostPresenter.java
package dev.ragnarok.fenrir.mvp.presenter;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.domain.IWallsRepository;
import dev.ragnarok.fenrir.domain.Repository;
import dev.ragnarok.fenrir.model.AttachmentEntry;
import dev.ragnarok.fenrir.model.Post;
import dev.ragnarok.fenrir.mvp.view.IProgressView;
import dev.ragnarok.fenrir.mvp.view.IRepostView;
import dev.ragnarok.fenrir.util.RxUtils;


public class RepostPresenter extends AbsAttachmentsEditPresenter<IRepostView> {

    private final Post post;
    private final Integer targetGroupId;
    private final IWallsRepository walls;
    private boolean publishingNow;

    public RepostPresenter(int accountId, Post post, Integer targetGroupId, @Nullable Bundle savedInstanceState) {
        super(accountId, savedInstanceState);
        walls = Repository.INSTANCE.getWalls();
        this.post = post;
        this.targetGroupId = targetGroupId;

        getData().add(new AttachmentEntry(false, post));
    }

    @Override
    public void onGuiCreated(@NonNull IRepostView viewHost) {
        super.onGuiCreated(viewHost);
        viewHost.setSupportedButtons(false, false, false, false, false, false);

        resolveProgressDialog();
    }

    private void resolveProgressDialog() {

        if (publishingNow) {
            callView(v -> v.displayProgressDialog(R.string.please_wait, R.string.publication, false));
        } else {
            callView(IProgressView::dismissProgressDialog);
        }
    }

    private void setPublishingNow(boolean publishingNow) {
        this.publishingNow = publishingNow;
        resolveProgressDialog();
    }

    private void onPublishError(Throwable throwable) {
        setPublishingNow(false);
        callView(v -> showError(v, throwable));
    }

    @SuppressWarnings("unused")
    private void onPublishComplete(Post post) {
        setPublishingNow(false);
        callView(IRepostView::goBack);
    }

    public final void fireReadyClick() {
        setPublishingNow(true);

        int accountId = getAccountId();
        String body = getTextBody();

        appendDisposable(walls.repost(accountId, post.getVkid(), post.getOwnerId(), targetGroupId, body)
                .compose(RxUtils.applySingleIOToMainSchedulers())
                .subscribe(this::onPublishComplete, this::onPublishError));
    }
}

File: app/src/main/kotlin/dev/ragnarok/fenrir/mvp/presenter/RepostPresenter.kt
package dev.ragnarok.fenrir.mvp.presenter

import android.os.Bundle
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.domain.IWallsRepository
import dev.ragnarok.fenrir.domain.Repository
import dev.ragnarok.fenrir.fromIOToMain
import dev.ragnarok.fenrir.model.AttachmentEntry
import dev.ragnarok.fenrir.model.Post
import dev.ragnarok.fenrir.mvp.view.IRepostView

class RepostPresenter(
    accountId: Int, private val post: Post,
    private val targetGroupId: Int?, savedInstanceState: Bundle?
) :
    AbsAttachmentsEditPresenter<IRepostView>(accountId, savedInstanceState) {
    private val walls: IWallsRepository = Repository.walls
    private var publishingNow = false
    override fun onGuiCreated(viewHost: IRepostView) {
        super.onGuiCreated(viewHost)
        viewHost.setSupportedButtons(
            photo = false,
            audio = false,
            video = false,
            doc = false,
            poll = false,
            timer = false
        )
        resolveProgressDialog()
    }

    private fun resolveProgressDialog() {
        if (publishingNow) {
            view?.displayProgressDialog(
                R.string.please_wait,
                R.string.publication,
                false
            )
        } else {
            view?.dismissProgressDialog()
        }
    }

    private fun setPublishingNow(publishingNow: Boolean) {
        this.publishingNow = publishingNow
        resolveProgressDialog()
    }

    private fun onPublishError(throwable: Throwable) {
        setPublishingNow(false)
        showError(throwable)
    }

    private fun onPublishComplete() {
        setPublishingNow(false)
        view?.goBack()
    }

    fun fireReadyClick() {
        setPublishingNow(true)
        val accountId = accountId
        val body = getTextBody()
        appendDisposable(walls.repost(accountId, post.vkid, post.ownerId, targetGroupId, body)
            .fromIOToMain()
            .subscribe({ onPublishComplete() }) { throwable ->
                onPublishError(
                    throwable
                )
            })
    }

    init {
        data.add(AttachmentEntry(false, post))
    }
}